### ISSUE TITLE
[Feature Request] Handle S3-compatible storage

### ISSUE THREAD

**Issue by dhess** [Dusk-Labs/dim#245]
One thing that's frustrating about all existing open source media management systems is the lack of S3-compatible storage options. I think it's especially ironic given that they all go out of their way to provide a nice Docker container, but then ignore the biggest (or maybe 2nd biggest) problem of the container approach, which is provisioning and accessing persistent, file-based storage.

Dim could really differentiate itself by providing first-class support for S3 interfaces:

* Scan, tag, & organize media on S3-compatible object storage systems.
* Provide the usual HTTP-based streaming APIs to clients for media located on S3-compatible object storage systems.

**Comment by vgarleanu** [comment]
Wouldn't it make more sense to have some sort of script that can mount a S3 bucket with fuse?

**Comment by k3d3** [comment]
This is usually handled by something like rclone, which FUSE-mounts an S3 bucket, then you point the media server at that FUSE mount.

**Comment by xd003** [comment]
Agreed on this being one of the very basic thing missing at the moment . i did setup everything up only to realise at the end that my rclone mounted google drive isn't being detected in library creation stage . More than half of the people i know use rclone mounted cloud storage for their mediaservers . I hope this gets added one a bit more prioritised basis . Looking forward to try it out once fuse mounts are supported

**Comment by vgarleanu** [comment]
@xd003 When you mount google drive with rclone, does the path simply not show up in the library creation modal? I'd be happy to prioritize this issue if I can get more details on your setup and environment.

**Comment by vgarleanu** [comment]
Seems like the scanner is stalling on something, Ill look into it, thanks for reporting.

**Comment by moistdio** [comment]
The same thing happens to me. 
`start_custom{library_id=3 media_type=Movie provider=TMDBMetadataProviderOf<K> { key: Movie }}: dim::scanner: Walked all target directories. elapsed_ms=89 files=287`

**Comment by grapemix** [comment]
We don't have to use object storage layer from the cloud provider, we can simply host by ourselves.
Object layer or s3-compatiable storage should be the next generation storage for homelab or production env because of:
- Build-in object tagging
- Build-in versioning in object level
- Third-party integrated backup solution
- Build-in ACL
- Redudency (not just in disk level, but also instance level)
- Easy horizontal scalability (not just in disk level, but also instance level)
- Easy size expansion (we don't have to copy files if we want to expand the storage)
- Able to share file in limited time (like via pre-signed url)
- Easy to find library to upload from client side (like via pre-signed url)

If we have truly object storage layer, the docker container can become truly stateless which means we can have multiple k8s pods and spin up and down if we need to among different instances. It is painful to share persistent volume between different pods/projects.... Yes, we can hack it, so I mean painful to watch ;)

Is it overkill? probably, but we can also simply use midnight commander to watch videos too if we are minimalist , right? ;)

**Comment by mental32** [comment]
Alright so Github sorting by least recently active isn't exactly accurate.

**Comment by malthe** [comment]
As @dhess points out, one of the main benefits of network-based storage is:

> Provide the usual HTTP-based streaming APIs to clients for media located on S3-compatible object storage systems.

That is, a client can pull blocks directly from the upstream source, at least when transcoding is not required. And if transcoding _is_ required, there are services such as [AWS Elastic Transcoder](https://aws.amazon.com/elastictranscoder/) that do just this.
